pong/hardware/lcd.c

```diff
@@ -0,0 +1,187 @@
+#include "lcd.h"
+#include "../utility/utils.h"
+#include <string.h>
+
+#define MAX_LEN 1024
+
+#define VALUE_FILE "value"
+
+static void initPin(char* fileDir, char* pin, char* pinNumber);
+
+static void enableWrite(LcdScreen screen);
+
+static void write4BitsToLcd(LcdScreen* screen, __uint8_t value);
+
+LcdScreen* generateLcd(LcdHardware lcdHardware) {
+    LcdScreen* newLcdScreen = malloc(sizeof(LcdScreen));
+    newLcdScreen->numDataPins = lcdHardware.numDataPins;
+    newLcdScreen->dataPinsFilePaths = calloc(lcdHardware.numDataPins, sizeof(char*));
+    newLcdScreen->ePinFilePath = calloc(1, sizeof(char) * MAX_LEN);
+    newLcdScreen->rsPinFilePath = calloc(1, sizeof(char) * MAX_LEN);
+    sprintf(newLcdScreen->ePinFilePath, "%s%s/", GPIO_FILE_DIR, lcdHardware.ePin.gpioPin);
+    initPin(newLcdScreen->ePinFilePath, lcdHardware.ePin.pin, lcdHardware.ePin.pinNumber);
+    
+    for (int i = 0; i < lcdHardware.numDataPins; i++) {
+        newLcdScreen->dataPinsFilePaths[i] = calloc(1, sizeof(char)*MAX_LEN);
+        sprintf(newLcdScreen->dataPinsFilePaths[i], "%s%s/", GPIO_FILE_DIR, lcdHardware.dataPins[i].gpioPin);
+        initPin(newLcdScreen->dataPinsFilePaths[i], lcdHardware.dataPins[i].pin, lcdHardware.dataPins[i].pinNumber);
+    }
+
+    sprintf(newLcdScreen->rsPinFilePath, "%s%s/", GPIO_FILE_DIR, lcdHardware.rsPin.gpioPin);
+    initPin(newLcdScreen->rsPinFilePath, lcdHardware.rsPin.pin, lcdHardware.rsPin.pinNumber);
+    
+    /**
+     * The following segment of code is adapted from 
+     * https://opencoursehub.cs.sfu.ca/mba200/grav-cms/ensc351/links/files/2022-student-howtos/16x2CharacterLCDThroughGPIO.pdf
+    */
+
+    // Special Function Set 1.
+	write4BitsToLcd(newLcdScreen, 0x03); // 0011
```

Comment:
QOL: The write4BitsToLcd function calls are indented by 1 more tab than they should be
fix: shift tab the relevant lines